README.md
# What is UliPad

Ulipad is a wxPython powered, programmer oriented and flexible editor.
It has many features such as class browser,
code auto-complete, html viewer, directory browser, wizard and many
others. The most interesting
and distinctive feature is the use of mixin technique which makes
Ulipad an easy-to-extend
programming environment. You can write your own mixins, plugins or
simple scripts, and all of
them can be integrated in Ulipad in a seamless way.

# Why named it UliPad?

The name comes from Claudio Grondi (thank you!).

Claudio Grondi said:

> As we are on the way to make NewEdit THE programmer editor I think, that
> a new unique name will help here much (NewEdit is just cheap...).
>
> I suggest to use 'UliPad'.
>
> It has the shortcut of UnLImited.
> It tells it is a plain text editor (Pad).
> It tells something about the original author (LImodou).
>
> But the most important advantage is:
>
>    Google has no entry for 'UliPad' yet!
>
> Wonderful! It seems many people like it.

# Description

UliPad uses Mixin and Plugin technique as its architecture. Most of its classes can be extended via mixin and plugin components, and finally become an integrity class when creating the instance. So `UliPad` is very dynamic. You can write the new features in new files, and hardly need to modify the existing code. And if you want to extend the existing classes, you could write mixins and plugins, and this will be bound to the target class that I call "Slot Class". This technique will make the changes centralized and easily managed.

Requirements:

 * Python 2.6+
 * wxPython 2.8+ Unicode Version

Some plugins may have additional requirements, please read the readme information before installing them.

# Objective

Make a clean, powerful, flexible general editor, and even an application framework.  And adding extensions should be very easy.

Tasks:

* Mixin and Plugin framework
* Unicode support
* User defined plugin management
* User custom window integration
* User custom wizard function

# Change Log

You can see the `ChangeLog` to know the new changes of UliPad.

# Features

* **Cross platform**
    * based on wxPython, so it can run anywhere that wxPython works, such as: Windows, Linux.
    * Unicode support.
* **Most features of wxStyledTextCtrl(Scintilla)**
    * Syntax highlighting, support Python, c/c++, html, plain text
    * Folding
    * Brace Matching
    * ...
* **Extended selection**
    * Extended word selection -- You can press Ctrl+`MouseDoubleClick` to select a word including '.'
    * Matched selection -- Select text in quoted chars like: (), [], {}, `''`, `""`.

        For example: a string just like

        ```
        def func(self, 'This is a test'):
                   ^
        ```

        The '^' char represents caret position in above line. If you press Ctrl+E, you will select the whole text in (),
        i.e. "self, 'This is a test'". Something more in Selection Menu.
* **Other editing extension**
    * Duplicating text -- Just like Vim Ctrl+V, Ctrl+P, and more. You can duplicate above or below char, word, line which match the leading chars.
    * Quoting text -- Add some quoted chars before and after selected text, just as: `""`, `''`, (), [], {}, and customized string, etc.
    * Text convertion and view -- python -> html, reStructured Text -> html, textile -> html, and you can output or view the html text in message window, or html view window, or replace the selected text.
    * Utf-8 encoding auto detect
    * Changing document encoding
    * Auto backup
    * Last session support -- It'll save all the filenames as closed, and reopen the files as next started.
    * Smart judge the indent char -- It'll auto guess the indent char, and sets it.
    * Finding in files
    * Bookmark supports
* **Python support**
    * built-in python interactive window based on `PyShell`, support Unicode
    * Auto completion
    * Function syntax calltips
    * Run, run with argument, stop python source
    * Auto change current path
    * Python class browser
    * Syntax and PEP8 style checking，also supply a pylint plugin.
* **Code snippets**

    You can manage your code snippets with categories, and each category can have many items. Every item will represent a code snippet. You can insert an item just by double-clicking on it. It even supports importing and exporting.

* **Simple project support**

    Can create a special file `_project`, so every file and folder under the folder which has the `_project` can be considered as a whole project.

* **Extension mechanism**
    * Script -- You can write easy script to manipulate the all resource of UliPad, just like: text conversion, etc.
    * Plugin -- Customized function. More complex but more powerful. Can easily merge with UliPad, and can be managed via menu.
    * Shell command -- Add often used shell commands, and execute them.
* **Ftp support**

    You can edit remote files through ftp. You can add, rename, delete, upload, download file/directory.

* **Multilanguage support**

    Currently supports 4 languages: English, Spanish, Simplified Chinese and Traditional Chinese, which can be auto-detected.

* **Ships many plugins** (must be configed as used them before)
    * Django support plugin
    * Batch rename files plugin
    * Collaborative Programming support plugin, names as *pairprog*.
    * Mp3 player plugin
    * Spell check plugin
    * wizard plugin
    * Text to speech(windows only) plugin
    * ...
* **Shipped scripts**
    * You can find them in (`$UliPadInstalled`)/scripts.
* **Wizard**

    You can make your own wizard template. The wizard can input user data, combine with template, and output the result. And wizard also support code framework created. This feature will help you improving coding efficiency.

* **Direcotry Browser**

    Browse multiple directories, and you can really add, delete, rename directories and files. Double click will open the file in Editor window.

* **AutoComPlete(acp)**

    Suport user autocomplete file, it can help to input code very helpful and functional.

* **Column Editing Mode**

    You can select multilines, and then set a column mode region, so in any line of this region, if you enter a character, other lines will also add this character. If you want to deal with multilines as a similar mode, this functionality will be very handy.

* **Smart Navigation**

    UliPad can remember the visit order of your opened files, and you can go back or go forward in  these files.

* **Live regular expression searching**

    You can type some regular expression on the fly, and see the result dynamiclly.

* **Spell check plugin**

    Need to install `PyEnchant` module.

* **Collaborative Programming**

    Multi-user can modify some files at the same time. You should enable *pairprog* plugin.

* **Todo Supports**

    Auto finds todos and supports several kind of formats.

* **Multi-View Supports**

    User can open a document in multi views, for example in left pane or bottom pane.

* **Version Control Support**
    * svn support. Now you can use svn in `UliPad` to update, checkout, commit, etc.


# How to run

Just download the source code and extract to any folder, and then execute:

```
python ulipad.pyw
```

or

```
python ulipad.py
```

# License

UliPad is released under GPL license.


ChangeLog.txt
## 2014/03/10

* Add markdown filetype and render support, you can just like rest filetype view the output in pane.
* Add git status display in directory browser window
* Fix Column Mode bug

[2013/10/15]
InputAssistant.py   Fix get_calltip bug.

[2011/04/21]
mPad.py             Add configurable memo file support. So you can set something like this:
                    
                    [default]
                    memo = "C:\\path\\to\\memo.txt"
scriptils.py        Add some useful methods to support scripts files. Such as emptytab, newtab, etc.
LexerClass1.py mLexer.py Add Bash support.

All above thanks to Helio Perroni Filho <xperroni@gmail.com>
[2010/04/22]
mGuessLang.py       Add #! /usr/bin/env python support thanks for zhangchunlin
plugin/ruby/mClassBrowser.py
                    Fix ruby class browser bug
Add Lua plugin      Thanks to zhangchunlin.

[2008/06/02]
mEncoding.py EncodingDialog.py
                    Migirate EncodingDialog from xrc to meide
LexerBase.py mLexerBase.py LexerClass.py
                    Add default color theme support, you can set it in Preference
                    
[2008/05/14]
mPythonProject.py   Add Create Python Package in context menu of Directory Browser Window

[2008/05/12]
RegexWindow.py      Improve regex window, and when you set Unicode flag, it'll 
                    automatically convert \uXXXX to unichr
[2008/05/05]
mEditorCtrl.py      Fix Cancel Save file will prompt save as dialog again bug

[2008/05/01]
mFormat.py          Fix strip tailing spaces bug

-------------------------------------- 4.0 -----------------------------
[2008/04/24]
Fix                 Cann't to Can't

[2008/04/19]
mInputAssistant.py  Add config menu, and add toggle input assistant menu item
DirBrowser.py       Fix refresh directory bug

[2008/04/18]
mWrapText.py wraptext.py
                    Add wrap text feature

[2008/04/15]
doc                 Modify the documents
mEditorCtrl.py      Add copy filename to clipboard menu on document area tab context menu

[2008/04/14]
mEditorCtrl.py mDirBrowser.py DireBrowser.py
                    Add Open Explorer Window Here menu to editoctrl tab
                    Move Close Other to Close menu group in notebook context menu
                    Add tool button snippet and modify open dirbrowser and snippet as check toolbutton
                    Change `explorer.exe %s` as `explorer.exe /e, %s`
ruler/__init__.py   Fix ruler bug
DirBrowser.py       When add path, it'll automatically pop up project setting dialog

[2008/04/12]
mLexer.py mLexerClass1.py
                    Add slice syntax support
mCheckUpdate.py     Add auto check update

[2008/04/10]
mCommands.py mConcurrent.py
                    Improve caret display(process KILL_FOCUS event if in command mode or concurrent mode)

[2008/04/09]
mFileNew.py         Can remember the last new file type(select new type menu first),
                    then when you select new menu, it'll use the last new file type as default
PrefDialog.py mInputAssistant.py
                    Improve preference dialog input assistant checkbox process.
                    When you check the first checkbox(Enable input assistant) it'll
                    automatically toggle other 5 checkboxes.

[2008/04/03]
SyntaxCheck.py      Fix exception process bug

[2008/03/31]
mLineending.py      Add line ending mixture check when saving file feature
FindReplace.py      Add search text count

[2008/03/27]
EditorFactory.py    Chanage left double click to right double click
common.py makemenu.py
                    Change setmenutext to use fix width to set the menu text, replace with '\t'
[2008/03/22]
web2py_plugin       Add web2py plugin

[2008/03/17]
HtmlPage.py mRestFileType.py Del acp/rst
                    Improve ReST document render, and fix the setfocus lost bug when
                    auto modified the html output, thanks a lot to ygao

[2008/03/15]
mPyRun.py           Add auto detect python interpreter in windows platform

[2008/03/13]
mScript.py          Fix script filename cannot be unicode(chinese) bug

[2008/03/04]
*                   Enhance on_modified process, improve the edit speed
mShell.py           Strip leading space when doing "Run in Shell"
mAutoCheck.py       Chanage auto file check to Editor on_set_focus event
SvnSettings.py      Fix password widget is not Password type widget bug
mFileNew.py         Fix template line-ending not match the default line-ending setting

[2008/03/03]
Editor.py mFormat.py mLineending.py
                    Fix press tab and convert EOF too slow problem

[2008/03/02]
EditorFactory.py FlatNotebook.py
                    Fix tab invisible bug after changing size or changing the page title
mLineending.py      Add strip line ending when saving functionality and strip line ending menu item
                    in Edit/Format.

[2008/03/01]
mLineending.py      Fix copy and paste lineending is not correct bug

[2008/02/16]
Editor.py mEditor.py FlatNotebook.py mInputAssistant.py
                    Fix smart tabs bug.

[2008/02/15]
Editor.py           Skip control keys on keydown event

[2008/01/18]
common.py           Fix webopen bug, cann't correctly deal with 'mailto:'
mDDESupport.py      Fix DDE bug, thanks to LP <liupengf12@gmail.com>

[2008/01/16]
plugins/SpellCheck.py  67598092
                    Refactor interface

[2008/01/15]
common.py           Add icon set theme support.

[2008/01/14]
PrefDialog.py mPreference.py
                    Change preference dialog from notebook to treebook
AsyncAction.py mInputAssistant.py
                    Refactor the AsyncAction class, and add get_timestep() method.
                    Reference to ygao's implemention
Editor.py, *        Fix if lines are folded, when goto hiding lines will no effect bug.

[2008/01/12]
DDE.py mDDESupport.py
                    Change DDE to asyncore and asynchat framework

[2008/01/11]
mako_plugin/dirbrowserext.py makotmp.acp

setup_inno.py
# A setup script showing how to extend py2exe.
#
# In this case, the py2exe command is subclassed to create an installation
# script for InnoSetup, which can be compiled with the InnoSetup compiler
# to a single file windows installer.
#
# By default, the installer will be created as dist\Output\setup.exe.

from distutils.core import setup
import py2exe
import os, sys
from glob import glob

################################################################
# A program using wxPython

# The manifest will be inserted as resource into test_wx.exe.  This
# gives the controls the Windows XP appearance (if run on XP ;-)
#
# Another option would be to store if in a file named
# test_wx.exe.manifest, and probably copy it with the data_files
# option.
#
manifest_template = '''
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<assembly xmlns="urn:schemas-microsoft-com:asm.v1" manifestVersion="1.0">
<assemblyIdentity
    version="5.0.0.0"
    processorArchitecture="x86"
    name="%(prog)s"
    type="win32"
/>
<description>%(prog)s Program</description>
<dependency>
    <dependentAssembly>
        <assemblyIdentity
            type="win32"
            name="Microsoft.Windows.Common-Controls"
            version="6.0.0.0"
            processorArchitecture="X86"
            publicKeyToken="6595b64144ccf1df"
            language="*"
        />
    </dependentAssembly>
</dependency>
</assembly>
'''

RT_MANIFEST = 24

################################################################
# arguments for the setup() call

ulipad_wx = dict(
    script = "UliPad.pyw",
    other_resources = [(RT_MANIFEST, 1, manifest_template % dict(prog="UliPad"))],
    dest_base = r"prog\ulipad_wx")

zipfile = r"lib\sharedlib.zip"

from distutils.file_util import copy_file
from distutils.dir_util import mkpath

def gather_data_files(module, base_dir):
    s = []
    mod = __import__(module, [], [], [''])
    path = mod.__path__[0]
    length = len(os.path.dirname(path)) + 1 #include os.sep
    
    for root, dirs, files in os.walk(path):
        for f in files:
            fname, ext = os.path.splitext(f)
            if ext not in ('.py', '.pyc', '.pyo'):
                filename = os.path.join(root, f)
                desfname = filename[length:]
                sfilename = os.path.join(base_dir, desfname)
                mkpath(os.path.dirname(sfilename))
                copy_file(filename, os.path.join(base_dir, desfname))
                s.append(desfname)
    return s

includes = ["modules.EasyGuider.*", "modules.meteor.*", "modules.*"]
packages = ["docutils"]
options = {
        "py2exe":
        {
                "compressed": 0,
                "optimize": 2,
                "includes": includes,
                "packages": packages,
        }
}

################################################################

class InnoScript:
    def __init__(self,
                 name,
                 lib_dir,
                 dist_dir,
                 windows_exe_files = [],
                 lib_files = [],
                 version = "1.0"):
        self.lib_dir = lib_dir
        self.dist_dir = dist_dir
        if not self.dist_dir[-1] in "\\/":
            self.dist_dir += "\\"
        self.name = name
        self.version = version
        self.windows_exe_files = [self.chop(p) for p in windows_exe_files]
        self.lib_files = [self.chop(p) for p in lib_files]

    def chop(self, pathname):
        assert pathname.startswith(self.dist_dir)
        return pathname[len(self.dist_dir):]

    def create(self, pathname="dist\\ulipad.iss"):
        self.pathname = pathname
        ofi = self.file = open(pathname, "w")
        print >> ofi, "[Setup]"
        print >> ofi, "AppName=%s" % self.name
        print >> ofi, "AppVerName=%s" % self.name + ' ' + self.version
        print >> ofi, "AppPublisher=Limodou"
        print >> ofi, "AppPublisherURL=http://code.google.com/p/ulipad/"
        print >> ofi, "AppSupportURL=http://hi.baidu.com/limodou"
        print >> ofi, "AppUpdatesURL=http://code.google.com/p/ulipad/"
        print >> ofi, "DefaultDirName={pf}\%s" % self.name
        print >> ofi, "DefaultGroupName=%s" % self.name
        print >> ofi, "AllowNoIcons=yes"
        print >> ofi, "LicenseFile=%s\dist\COPYLEFT.txt" % os.getcwd()
        print >> ofi, "InfoBeforeFile=%s\dist\INFO.txt" % os.getcwd()
        print >> ofi, "Compression=lzma/max"
        print >> ofi, "SolidCompression=yes"
        print >> ofi
        print >> ofi, "[Tasks]"
        print >> ofi, 'Name: "desktopicon"; Description: "{cm:CreateDesktopIcon}"; GroupDescription: "{cm:AdditionalIcons}"; Flags: unchecked'
        print >> ofi, 'Name: "quicklaunchicon"; Description: "{cm:CreateQuickLaunchIcon}"; GroupDescription: "{cm:AdditionalIcons}"; Flags: unchecked'
        print >> ofi
        print >> ofi, "[Files]"
        print >> ofi, 'Source: "%s.exe"; DestDir: "{app}"; Flags: ignoreversion' % self.name
        print >> ofi, '; NOTE: Don\'t use "Flags: ignoreversion" on any shared system files'
        print >> ofi
        print >> ofi, "[Icons]"
        print >> ofi, 'Name: "{group}\%s"; Filename: "{app}\%s.exe"' % (self.name, self.name)
        print >> ofi, 'Name: "{group}\{cm:UninstallProgram,%s}"; Filename: "{uninstallexe}"' % self.name
        print >> ofi, 'Name: "{userdesktop}\%s"; Filename: "{app}\%s.exe"; Tasks: desktopicon' % (self.name, self.name)
        print >> ofi, 'Name: "{userappdata}\Microsoft\Internet Explorer\Quick Launch\%s"; Filename: "{app}\%s.exe"; Tasks: quicklaunchicon' % (self.name, self.name)
        print >> ofi
#        print >> ofi, "[Registry]"
#        print >> ofi, 'Root: HKCR; Subkey: "*\shell\UliPad"; Flags: uninsdeletekey deletekey;'
#        print >> ofi, 'Root: HKCR; Subkey: "*\\shell\\UliPad\\command"; Flags: uninsdeletekey deletekey; ValueType: string; ValueData: "{app}\\%s.exe \"\"%%L\"\""' % self.name
#        print >> ofi
        print >> ofi, "[Run]"
        print >> ofi, 'Filename: "{app}\%s.exe"; Description: "{cm:LaunchProgram,%s}"; Flags: nowait postinstall skipifsilent' % (self.name, self.name)
        print >> ofi
        print >> ofi, r"[Files]"
        for path in self.windows_exe_files + self.lib_files:
            print >> ofi, r'Source: "%s"; DestDir: "{app}\%s"; Flags: ignoreversion' % (path, os.path.dirname(path))
        for d in dirs:
            print >> ofi, r'Source: "%s\*"; DestDir: "{app}\%s"; Flags: recursesubdirs' % (d, d)
        print >> ofi

        print >> ofi, r"[Icons]"
        for path in self.windows_exe_files:
            print >> ofi, r'Name: "{group}\%s"; Filename: "{app}\%s"' % \
                  (self.name, path)
        print >> ofi, 'Name: "{group}\Uninstall %s"; Filename: "{uninstallexe}"' % self.name

    def compile(self):
        try:
            import ctypes
        except ImportError:
            try:
                import win32api
            except ImportError:
                import os
                os.startfile(self.pathname)
            else:
                print "Ok, using win32api."
                win32api.ShellExecute(0, "compile",
                                    self.pathname,
                                    None,
                                    None,
                                    0)
        else:
            print "Cool, you have ctypes installed."
            res = ctypes.windll.shell32.ShellExecuteA(0, "compile",
                                  self.pathname,
                                  None,
                                  None,
                                  0)
            if res < 32:
                raise RuntimeError, "ShellExecute failed, error %d" % res


################################################################

def runother():
    import os

    f = file('run_setup.bat', 'w')
    print >> f, 'cd dist/lib'
#    print >> f, '7z x sharedlib.zip -osharedlib'
#    print >> f, 'del sharedlib.zip'
#    print >> f, 'cd sharedlib'
#    print >> f, '7z a -tzip -mx9 ..\sharedlib.zip -r'
#    print >> f, 'cd ..'
#    print >> f, 'rd sharedlib /s /q'
    print >> f, 'upx --best *.*'
    print >> f, 'cd ../..'
    f.close()
    os.system('run_setup.bat')

from modules.Version import version

from py2exe.build_exe import py2exe

class build_installer(py2exe):
    # This class first builds the exe file(s), then creates a Windows installer.
    # You need InnoSetup for it.
    def run(self):
        # First, let py2exe do it's work.
        py2exe.run(self)

        lib_dir = self.lib_dir
        dist_dir = self.dist_dir

        # create the Installer, using the files py2exe has created.
        script = InnoScript("UliPad",
                            lib_dir,
                            dist_dir,
                            self.windows_exe_files,
                            self.lib_files, version)
        print "*** creating the inno setup script***"
        script.create()
        print "*** compiling the inno setup script***"
        script.compile()
        
        # Note: By default the final setup.exe will be in an Output subdirectory.
    def make_lib_archive(self, zip_filename, base_dir, files,
                         verbose=0, dry_run=0):
        datafiles = gather_data_files("docutils", base_dir)
        files.extend(datafiles)
        return py2exe.make_lib_archive(self, zip_filename, base_dir, files, verbose, dry_run)

################################################################

import shutil
def my_copytree(src, dst):
    """Recursively copy a directory tree using copy2().

    Modified from shutil.copytree

    """
    base = os.path.basename(src)
    dst = os.path.join(dst, base)
    names = os.listdir(src)
    if not os.path.exists(dst):
        os.makedirs(dst)
    for name in names:
        srcname = os.path.join(src, name)
        try:
            if os.path.isdir(srcname):
                my_copytree(srcname, dst)
            else:
                shutil.copy2(srcname, dst)
        except:
#            error.traceback()
            raise

dirs = ['acp', 'plugins', 'wizard', 'scripts', 'lang', 
    'doc', 'packages', 'images', 'tools', 'conf', 'resources']
for d in dirs:
    my_copytree(d, 'dist')

data_files = [
        ('', ['ulipad.ico', 'COPYLEFT.txt', 'INFO.txt', 'UliPad.exe.manifest']),
]

if sys.winver >= '2.6':
    data_files.append(("", glob(r'Microsoft.VC90.CRT\*.*')))
    
setup(
        version = version,
        description = "UliPad",
        name = "UliPad",
        author = "limodou",
        author_email="limodou@gmail.com",
        url="http://code.google.com/p/ulipad/",

        options = options,
        # The lib directory contains everything except the executables and the python dll.
        zipfile = zipfile,
        windows = [
                {
                        "script":"UliPad.pyw",
                        "icon_resources": [(1, "ulipad.ico")]
                }
        ],
        data_files = data_files,
    # use out build_installer class as extended py2exe build command
    cmdclass = {"py2exe": build_installer},
    )

